### PR TITLE
Implemented password hashing, SecurityFilterChain and roles.

### PR DESCRIPTION
**Pull request by sovisou**

### CODE REVIEW

File: src/main/java/com/onlinebookstore/controller/BookController.java [hunk 1]
```diff
@@ -30,38 +31,44 @@ public class BookController {
...
     @Operation(summary = "Get book by id", description = "Get a book by its identifier")
+    @PreAuthorize("hasRole('ROLE_USER')")
     public BookDto findById(@PathVariable Long id) {
         return bookService.findById(id);
     }
 
     @PostMapping
     @Operation(summary = "Create a new book", description = "Create a new book")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
     public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
```
Comment (JJJazl) [inline src/main/java/com/onlinebookstore/controller/BookController.java:50]: Please, add an appropriate status code for POST method

File: src/main/java/com/onlinebookstore/model/Role.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    private RoleName role;
+
+    @Override
+    public String getAuthority() {
+        return role.name();
+    }
+
+    public enum RoleName {
+        USER,
+        ADMIN
```
Comment (JJJazl) [inline src/main/java/com/onlinebookstore/model/Role.java]: Spring Security requires use a prefix ROLE_. You can add this prefix to your enum values or change method `getAuthority()` to return "ROLE_" + roleName.name(), but it's better just change enum values
```suggestion
        ROLE_USER,
        ROLE_ADMIN
```

File: src/main/java/com/onlinebookstore/repository/role/RoleRepository.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.onlinebookstore.repository.role;
+
+import com.onlinebookstore.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<User, Long> {
```
Comment (JJJazl) [inline src/main/java/com/onlinebookstore/repository/role/RoleRepository.java]:
```suggestion
public interface RoleRepository extends JpaRepository<Role, Long> {
```

File: src/main/java/com/onlinebookstore/security/CustomUserDetailsService.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(
+                        () -> new UsernameNotFoundException("Can't find user by email"));
+    }
```
Comment (JJJazl) [inline src/main/java/com/onlinebookstore/security/CustomUserDetailsService.java]:
```suggestion
        return userRepository.findByEmail(email)
                .orElseThrow(
                        () -> new UsernameNotFoundException("Can't find user by email: " + email));
    }
```

**Review by JJJazl — CHANGES_REQUESTED** [review]

File: src/main/java/com/onlinebookstore/config/SecurityConfig.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/auth/***", "/swagger-ui/**", "/v3/api-docs/**")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/config/SecurityConfig.java]:
```suggestion
                                .requestMatchers("/auth/**", "/swagger-ui/**", "/v3/api-docs/**")
```

File: src/main/java/com/onlinebookstore/model/User.java [hunk 6]
```diff
@@ -5,15 +5,23 @@
...
 import lombok.Setter;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
 
 @Entity
 @Table(name = "users")
 @Getter
 @Setter
-public class User {
+public class User implements UserDetails {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/model/User.java:28]: Let’s realize soft-deleted for the user entity and override the isEnabled() method.

File: src/main/resources/db/changelog/changes/03-create-roles-table.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: role
+                  type: enum('USER', 'ADMIN')
+                  constraints:
+                    nullable: false
+                    unique: true
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/03-create-roles-table.yaml]: add one empty line at the end of the file  (check other files)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix one important requirement ;)

File: src/main/java/com/onlinebookstore/model/User.java (hunk 6 above)
Comment (liliia-ponomarenko) [inline src/main/java/com/onlinebookstore/model/User.java:28]:
```suggestion
@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id = ?")
@Where(clause = "is_deleted=false")
public class User implements UserDetails {
```
Let’s complete soft-deleted realization ;)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]

File: src/main/java/com/onlinebookstore/config/SecurityConfig.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.config.http.SessionCreationPolicy;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@EnableWebSecurity
+@EnableMethodSecurity
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/config/SecurityConfig.java]: Do you really need both of these annotations?

File: src/main/java/com/onlinebookstore/model/Role.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/model/Role.java:15]:
```suggestion
@Entity
@Getter
@Setter
```

File: src/main/java/com/onlinebookstore/repository/user/UserRepository.java [hunk 10]
```diff
@@ -1,8 +1,11 @@
 package com.onlinebookstore.repository.user;
 
 import com.onlinebookstore.model.User;
+import java.util.Optional;
 import org.springframework.data.jpa.repository.JpaRepository;
 
 public interface UserRepository extends JpaRepository<User, Long> {
     boolean existsByEmail(String email);
+
+    Optional<User> findByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/repository/user/UserRepository.java:12]:
```suggestion
    @EntityGraph(attributePaths = "roles")
    Optional<User> findByEmail(String email);
```

File: src/main/java/com/onlinebookstore/security/CustomUserDetailsService.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(
+                        () -> new UsernameNotFoundException("Can't find user by email: " + email));
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/security/CustomUserDetailsService.java]:
```suggestion
                        () -> new EntityNotFoundException("Can't find user by email: " + email));
```

File: src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -24,6 +26,7 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
                     + requestDto.getEmail());
         }
         User user = userMapper.toModel(requestDto);
+        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
```
Comment (Elena-Bruyako) [inline src/main/java/com/onlinebookstore/service/impl/UserServiceImpl.java:34]: find role user and set it up
```suggestion
        user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
        user.setRoles(Set.of(role));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Pull request merged**